### PR TITLE
[Perf] Dedicated sync streams

### PR DESCRIPTION
**Pull request by ljedrz**
This PR is a draft implementation of dedicated sync streams listed in the [node hardening issue](https://github.com/ProvableHQ/snarkOS/issues/4140), as an alternative to the [general chunking proposal](https://github.com/ProvableHQ/snarkOS/pull/4055). Initially, it applies only to the validators, which will benefit from it the most.

The main reasons for the proposal are:
- increasing the performance of consensus (by reducing Head-of-Line blocking caused by large network messages)
- reducing DoS surface (by greatly reducing the maximum network message size)

Benefits over chunking:
- simpler non-sync network plumbing
- the potential to reduce the maximum network message size to a much greater degree

The general approach is as follows: the `Gateway`'s `Event` gains 2 new variants (in order to temporarily maintain backward compatibility via the existing `BlockRequest` and `BlockResponse`), `SyncRequest` and `SyncResponse`. When a node receives a `SyncRequest`, it responds with an address to a dedicated sync stream, and a short-lived access token that must be used in order to establish the connection. Once established, the responder sends `BlockResponse` messages (through the dedicated stream), and the existing `BlockSync` plumbing handles the rest. Once the maximum number of responses (if such a limit is desired) per a sync stream is sent, a new sync stream needs to be opened in order to receive more blocks.

The rough list of code changes, enumerated for simpler referencing if need be:
1. Some of the network messages are moved to `snarkos-node-network` to avoid circular dependencies (it will also make sense for the currently `Gateway`-only messages to reside there once this syncing is extended to non-validators).
2. New structs, `SyncToken` and `SyncResponse`, are introduced.
3. The `Event` is extended with 2 new variants, `SyncRequest` (holding a `BlockRequest`) and `SyncResponse` (which holds the new `SyncResponse` struct).
4. A `SyncStreams` object is introduced; it is essentially a node that requires no special peer handling or address resolution, and has a trivial handshake. Just like the `Gateway`, it contains a clone of sync-related `Sender`s, and the `LedgerService`. Using a node makes stream handling a lot simpler, and - compared to ad-hoc streams - reduces potential NAT/firewall issues (since only a single listener port is involved). The `Tcp` node is very lightweight, and so are its connections.
5. The `BlockSync` plumbing is adjusted to account for the new logic.

The current state of the PR: nodes can successfully establish dedicated sync streams and send/receive blocks, but I'm running into design details of the current `BlockSync` setup that are incompatible with the new approach; the problematic spots can be seen commented out in `block_sync.rs`. The issues I've identified thus far are:
- [ ] the sync requests are matched with addresses associated with `BlockLocators` (as opposed to the dedicated sync streams)
- [ ] the sync responses are currently distributed in a somewhat "fanout" fashion; instead, we should now also be prepared to send many `BlockResponse` (or even just `Block`) messages to single peers via a single stream; the expected block ranges should also be much larger, so as to minimize the number of `SyncRequest` messages that need to be handled by the `Gateway`

Once these are resolved, the related tests will also need some adjustment.

@kaimast please let me know if you have ideas on how the aforementioned `BlockSync` integration issues can be solved while maintaining backward compatibility, or suggest how this logic could be delegated elsewhere; feel free to commit to this PR if you'd like to integrate these changes with `BlockSync` in a way that's aligned with your design and future plans.

Open questions:
1. Do we want to "cycle" through several streams while syncing? It might be unnecessary, since at that point we're already past twofold authentication (the validator handshake + the access token). Since the sync streams are not used for anything else and are lightweight, I see little harm in it.
2. Should we limit the requestor to a single sync stream? This is a problem if we wanted to have high syncing performance while cycling streams (as we couldn't begin a new stream before concluding the existing one).
3. Do we want to limit the number of requested blocks? This needs to be weighed with the desired syncing redundancy factor and the performance implications it has for the providers. Note: dedicated sync streams are by design more robust than singular requests for blocks, so we may not need as much redundancy anymore.
4. The values for some of the `const`s.

### CODE REVIEW

File: node/network/src/sync_token.rs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+use rand::RngExt;
+use std::{
+    fmt,
+    io::{Read, Write},
+    ops::Deref,
+};
+
+#[derive(Clone, PartialEq, Eq, Hash)]
+pub struct SyncToken([u8; 32]);
```
Comment (vicsn) [inline node/network/src/sync_token.rs:27]: Can you document what this is and how it works? Are we exposed to worse MITM attacks compared to the Gateway handshake?

File: node/sync/src/node.rs [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+impl<N: Network> Disconnect for SyncStreams<N> {
+    async fn handle_disconnect(&self, peer_addr: SocketAddr) {
+        self.active_streams.lock().remove(&peer_addr);
+    }
+}
+
+#[async_trait]
+impl<N: Network> OnConnect for SyncStreams<N> {
+    async fn on_connect(&self, peer_addr: SocketAddr) {
+        // Check if we're the ones who provide the sync.
```
Comment (vicsn) [inline node/sync/src/node.rs:296]: Can two peers sync from each other concurrently? Should we add a test for these edge cases?

**Review by vicsn — COMMENTED** [review]
1./2. I would try and optimize for simplicity of the implementation. With the foundations in place, in the future we can more formally balance robustness and performance under various scenario's.

3./4. Can you take a first stab in a google doc for the limits we choose? We can try to allocate, say, a *rough* 10GB at any given time for all peers combined. Assuming current average blocksize if needed.

File: node/network/src/sync_token.rs (hunk 1 above)
Comment (ljedrz) [reply node/network/src/sync_token.rs:27]: The token is sent in plaintext and could be observed or tampered with by an inline MITM, yes. However, it is strictly an anti-DoS mechanism to prevent unauthorized resource consumption on the validator, not a cryptographic session key. The integrity of the downloaded blocks is guaranteed by the hashes and signatures, not by the transport layer, so any MITM tampering would be instantly caught and rejected.

File: node/sync/src/node.rs (hunk 2 above)
Comment (ljedrz) [reply node/sync/src/node.rs:296]: This behavior is controlled by the `BlockSync` logic, and remains unchanged. I don't see how such a scenario would be useful, so I'm pretty sure it is disallowed. As for the comment, this check is only required due to the modular nature of the `Tcp` plumbing - at the point of `OnConnect::on_connect`, we don't readily know the side of the connection (though this could be exposed by the `Tcp` if needed), so we can look up the applicable block request (which we would need anyway) in order to check it.

File: node/network/src/block_response.rs [hunk 3]
```diff
@@ -128,12 +116,12 @@ impl<N: Network> DataBlocks<N> {
...
-        let candidate_end_height = 1 + self.last().map(|b| b.height()).unwrap_or(0);
+        // let candidate_start_height = self.first().map(|b| b.height()).unwrap_or(0);
+        // let candidate_end_height = 1 + self.last().map(|b| b.height()).unwrap_or(0);
         // Check that the range matches the block request.
-        if start_height != candidate_start_height || end_height != candidate_end_height {
-            bail!("Peer '{peer_ip}' sent an invalid block response (range does not match block request)")
-        }
+        // if start_height != candidate_start_height || end_height != candidate_end_height {
+        //     bail!("Peer '{peer_ip}' sent an invalid block response (range does not match block request)")
+        // }
```
Comment (ljedrz) [inline node/network/src/block_response.rs:124]: if we are to reuse the existing `BlockRequest`, this check is no longer correct, as the request can span a much greater range of blocks than the responses

File: cli/src/commands/start.rs [hunk 4]
```diff
@@ -155,6 +155,10 @@ pub struct Start {
     #[clap(long, requires = "validator")]
     pub bft: Option<SocketAddr>,
 
+    /// Set the IP address and port used for providing block synchronization.
```
Comment (kaimast) [inline cli/src/commands/start.rs:158]: Should we add a comment here `Will be set to 0.0.0.0:6130 if not specified`?

File: cli/src/commands/start.rs [hunk 5]
```diff
@@ -507,6 +511,13 @@ impl Start {
             self.node = Some(address);
         }
 
+        if self.sync_listener.is_none() {
+            let port = get_devnet_router_address_for_node(dev).port() + 2000; // 6130+
```
Comment (kaimast) [inline cli/src/commands/start.rs]: Can we add a `get_devent_sync_address_for_node` method?

File: cli/src/commands/start.rs [hunk 6]
```diff
@@ -846,7 +860,7 @@ impl Start {
 
         // Initialize the node.
         let node = match node_type {
-            NodeType::Validator => Node::new_validator(node_ip, self.bft, rest_ip, self.rest_rps, account, &trusted_peers, &trusted_validators, genesis, cdn, storage_mode, node_data_dir, self.trusted_peers_only, self.auto_db_checkpoints.clone(), dev_txs, self.dev, signal_handler.clone()).await,
+            NodeType::Validator => Node::new_validator(node_ip, sync_listener, self.bft, rest_ip, self.rest_rps, account, &trusted_peers, &trusted_validators, genesis, cdn, storage_mode, node_data_dir, self.trusted_peers_only, self.auto_db_checkpoints.clone(), dev_txs, self.dev, signal_handler.clone()).await,
```
Comment (kaimast) [inline cli/src/commands/start.rs:864]: We should open an issue about improving how `Node` is created. The number of arguments is getting a little out of hand.
A builder pattern or a struct of arguments would be better at this point.

File: node/network/src/block_request.rs [hunk 7]
```diff
@@ -38,20 +40,6 @@ impl BlockRequest {
     }
 }
 
-impl EventTrait for BlockRequest {
```
Comment (kaimast) [inline node/network/src/block_request.rs:41]: Could we have it implement `Display` and then use the display trait downstream?

File: node/bft/events/src/lib.rs [hunk 8]
```diff
@@ -90,6 +83,52 @@ pub trait EventTrait: ToBytes + FromBytes {
...
+        }
+        .into()
+    }
+}
+
+impl<N: Network> EventTrait for CertificateRequest<N> {
+    /// Returns the event name.
+    #[inline]
+    fn name(&self) -> Cow<'static, str> {
+        "CertificateRequest".into()
```
Comment (kaimast) [inline node/bft/events/src/lib.rs:120]: Could this print the certificate ID as well? The round number would also help, but it is not included in the message (I think).

File: node/sync/src/helpers/sync_channel.rs [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+
+use anyhow::{Result, anyhow};
+use std::net::SocketAddr;
+use tokio::sync::{mpsc, oneshot};
+
+const MAX_CHANNEL_SIZE: usize = 8192;
+
+#[derive(Clone, Debug)]
+pub struct SyncSender<N: Network> {
+    pub tx_block_sync_insert_block_response: mpsc::Sender<(
```
Comment (kaimast) [inline node/sync/src/helpers/sync_channel.rs]: Could some of these be private? There are helper functions for some of these that should be used instead of directly accessing them.

File: node/sync/src/node.rs (hunk 2 above)
Comment (kaimast) [reply node/sync/src/node.rs:296]: I think it could happen if for some reason a block request gets delayed. However, it is very unlikely.

**Comment by kaimast** [comment]
I left a bunch of nitpicky comments, but overall the PR looks great so far!

File: node/network/src/block_request.rs (hunk 7 above)
Comment (ljedrz) [reply node/network/src/block_request.rs:41]: totally, pretty sure it's an artifact of some old development plans that were ultimately abandoned

Comment (ljedrz) [reply node/network/src/block_request.rs:41]: opened a dedicated issue: https://github.com/ProvableHQ/snarkOS/issues/4243

File: node/bft/events/src/lib.rs (hunk 8 above)
Comment (ljedrz) [reply node/bft/events/src/lib.rs:120]: [added](https://github.com/ProvableHQ/snarkOS/issues/4243#issuecomment-4350608258) to the dedicated `fmt::Display` issue

File: cli/src/commands/start.rs (hunk 6 above)
Comment (ljedrz) [reply cli/src/commands/start.rs:864]: created a dedicated issue: https://github.com/ProvableHQ/snarkOS/issues/4244

File: node/sync/src/helpers/sync_channel.rs (hunk 9 above)
Comment (ljedrz) [reply node/sync/src/helpers/sync_channel.rs]: this is just a 1:1 move, but yes, it seems that 2 of these members can be made private right now

**Comment by ljedrz** [comment]
@kaimast thanks for the comments so far; also highlighting the request from the description, as I wouldn't want to break the current setup due to its incompatibilities with the new design:
> please let me know if you have ideas on how the aforementioned `BlockSync` integration issues can be solved while maintaining backward compatibility, or suggest how this logic could be delegated elsewhere; feel free to commit to this PR if you'd like to integrate these changes with `BlockSync` in a way that's aligned with your design and future plans.

**Comment by ljedrz** [comment]
Rebased to fix a conflict.

File: node/network/src/block_response.rs (hunk 3 above)
Comment (kaimast) [reply node/network/src/block_response.rs:124]: Wasn't the plan to change block requests to individual blocks?

Comment (ljedrz) [reply node/network/src/block_response.rs:124]: yes, ultimately I'd want to do so, but I wanted to minimize the diff, at least for now; this would also be easier once the backward compatibility is no longer needed

**Comment by kaimast** [comment]
Does this branch already work? It would be useful to see whether the performance regression noted in #4224 is also present here. 

I am not sure if you have permission to trigger the benchmark workflow. If you do, please trigger when the initial implementation is done.

**Comment by ljedrz** [comment]
No, as I'm not sure how to plug these changes into `BlockSync` while maintaining backward compatibility - I was hoping you could suggest how to work around the locator address matching and range limits.

**Comment by vicsn** [comment]
@ljedrz can you update the precise benefit of this approach in the README above after https://github.com/ProvableHQ/snarkOS/pull/4419 is merged

**Comment by cbeck88** [comment]
I think the main thing is 4419 doesn't help the gateway, it only helps the router.

So it doesn't help for threat model of a validator node that goes Byzantine and tries to DOS the other nodes.
